text/0000-let-expression.md

```diff
@@ -0,0 +1,628 @@
+- Feature Name: let_expression
+- Start Date: 2021-08-08
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Convert `let <pat> = <expr>` from a statement to an expression of type `bool`. Make
+the language more consistent and remove corners by generalizing, plus adding
+a bunch of magic constructs that are useful. After this RFC, you'll be able to write, among many other things:
+
+```rust
+// generalized matches! macro:
+assert!((let Some(x) = y) && x > 2);
+
+// generalized let-else construct
+(let Some(a) = b) && (let Some(c) = f(&a)) || {
+    return Err("failed");
+};
+println!("{}, {}", a, c);
+
+// generalized assignment with default
+(let Some(x) = y) || (let Foo(x) = bar) || (let x = default);
+println!("{}", x);
+```
+
+# Motivation
+[motivation]: #motivation
+
+The main motivation for this RFC is improving consistency and ergonomics of language.
+
+## Consistency
+
+Currently we have `if`, `if let` and `if let && let` and we teach them as three different
+constructs (plus `let else` in future). But if we make `let` a `bool` expression, all will become the same and it would be
```

Comment:
The reasoning here is, new learners when see `if let`, expect that `let` should behave like a bool expression (like anything else inside `if` scrutinee) and this confusion will grow if we chain them with `&&` in if-let-chain and `||` in let-else. By making them an actual `bool` expression, there will be more things to learn (about PB and NB), but there would be no surprise for how `let` statement jumped into `if` and `&&` because it is a `bool` expression.

Reply:
Do new learners actually expect `let` to be a bool expression? The issue you linked in the RFC did not have any mention of bools. And new learners will get confused by the concepts of PBs and NBs too. You clearly have a bias towards if-let-chain and let-else because you describe their syntax as confusing, but as "more things to learn" when you talk about your RFC.

Reply:
Yah, the linked issue was renamed and labeled as a bug in error reporting, the error wasn't helpful to the end user and had nothing really to do with booleans at all